Move Paraglob To C++ & Update Data Strucures

0xekez opened the pull request:
The current paraglob implementation has a really nice algorithm but is slowed down by its internal set data structure. This is discussed some [here](https://github.com/zeek/zeek/issues/268). Quoted from the link:

_With the old paraglob adding a set of 5,000 patterns took 7.5 seconds and then it could perform about 6,500 queries a second on those patterns. The time to build the paraglob also grew in N^2 time. Building a paraglob with 10,000 patterns took 55.5 seconds._

_This version can add a set of 5,000 patterns in 0.66 seconds and the time taken grows lineally with the number of patterns. Adding 10,000 patterns takes about 1.3 seconds. It also performs queries much much faster than the old paraglob, starting at about 2x as fast for small pattern sets and widening the gap as the size of the pattern set increases. This varies with the percentage of patterns that match in the dataset. It is faster when a higher percentage of patterns match._

This commit moves all of paraglob's source code to C++ except for the aho-corasick trie which it adds a C++ interface to. It also uses an `unordered_map` from the C++ standard library as a replacement for the set which leads to a substantial speed-up both for making the paraglob and running queries on it.

Let me know if there are any changes you would like to see in this and I'll get to them ASAP. I'm still a little shaky on what sort of boilerplate is needed for something to be in a zeek repo, but I'm happy to add anything that is missing. I'd love any code review/critique anyone has as well. The commit is really long looking, but the most notable files are `paraglob.h/cpp` and `paraglobNode.h/cpp`.

Thanks a bunch.

Edit: I made an error squashing some of my commits, but fixed up the problem.

0xekez committed:
Move to C++ & update data structures

0xekez committed:
Move to vector internally.

Earlier use of set works, but appeared to be much faster than it
actually was. std::set::merge was deleting the contents of the set
being merged into the pattern set. As a result, after a while all
of the pattern sets would be deleted and there would be no patterns
to match against == looks like good performance.

rsmmr commented on AhoCorasickPlus.cpp:
```diff
@@ -0,0 +1,125 @@
+/*
+ * AhoCorasickPlus.cpp: A sample C++ wrapper for Aho-Corasick C library
```
As this file is part of multifast as well, move into `ahocorasick`?

rsmmr commented on AhoCorasickPlus.h:
```diff
@@ -0,0 +1,76 @@
+/*
```
Same here, move into ahocorasick.

rsmmr commented on CMakeLists.txt:
```diff
@@ -1,6 +1,16 @@
-cmake_minimum_required (VERSION 2.8)
+cmake_minimum_required (VERSION 3.14)
```
This should build and install a paragon library, plus the headers; but probably then doesn't need to install the driver as most people will never use it. 

Once we pull this into Zeek via submodule, it'll need some more tweaking to fit into its CMake setup.

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+
+Paraglob::Paraglob(const std::vector<std::string> &patterns): Paraglob() {
+  for (std::string pattern : patterns) {
+    this->add(pattern);
```
`this->` seems like an unusual style, any particular reason for using that?

0xekez replied:
This is just a stylistic thing that I've picked up over time. I tend to think that for functions with generic names like add its better to be explicit that I'm calling the classes method. In my personal opinion as well I think code like this (while a little more verbose) is a little more clear for reading. If you think its excessively verbose though I've got no problem getting rid of it :)

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+  if (m_words.size() == 0) {
+    this->single_wildcards.push_back(pattern);
+  }
+  for (std::string meta_word : m_words) {
```
Matter of taste, but generally we tend to use more white space for Zeek code, including in particular some empty lines between logical blocks to separate them visually. Also, single-line blocks don't really need the curly braces.

rsmmr commented on paraglob.h:
```diff
@@ -1,39 +1,39 @@
+
+
+class Paraglob {
+private:
```
I'd move the `private` parts to the end, that makes it easier to quickly see the class' API.

rsmmr commented on paraglobNode.cpp:
```diff
@@ -0,0 +1,22 @@
+#include "paraglobNode.h"
```
These methods are small enough that I think they could just move inline into the header. In fact, maybe move all the `ParaglobNode` stuff into the main `paraglob` header? Then it's just a single header file that's needed to use the library.

I'd also move everything internal (ParaglobNode) into an internal namespace to make clear it's not part of the public API.

rsmmr commented on paraglob.h:
```diff
@@ -1,39 +1,39 @@
+#include <algorithm> //remove_if, sort
+
+
+class Paraglob {
```
would be good to put this into its own namespace (`paraglob` I guess).

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+#include "paraglob.h"
+
+Paraglob::Paraglob(const std::vector<std::string> &patterns): Paraglob() {
+  for (std::string pattern : patterns) {
```
for (const auto& pattern : patterns)

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+    status = this->my_ac.addPattern(meta_word, patId);
+    if (status != AhoCorasickPlus::RETURNSTATUS_SUCCESS &&
+      status != AhoCorasickPlus::RETURNSTATUS_DUPLICATE_PATTERN) {
+      std::cout << "Failed to add: " << meta_word << std::endl;
```
Should return an error indicator instead of printing to stdout, so that the caller can handle as appropriate.

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+  // Narrow to the meta-word matches
+  AhoCorasickPlus::Match aMatch;
+  this->my_ac.search(text, false);
+  std::vector<std::string> meta_matches;
```
Do we need this temporary here? Why not build up `patterns` directly?

rsmmr commented on paraglob.cpp:
```diff
@@ -0,0 +1,111 @@
+  // Get the relevant patterns
+  std::vector<std::string> patterns;
+  for (std::string meta_word : meta_matches) {
+    this->meta_to_node_map.at(meta_word).merge_into(patterns);
```
Could we run fnmatch() on the elements right here and avoid having to remove them later?

0xekez replied:
Yes! That's clever. Doing this sped it up quite a bit.

rsmmr reviewed (commented):
- I suggest we start putting a bit a bit more structure in place for the files: say about `src/` directory for the main code, and `tools` for the driver and benchmark. Makes it easier to understand what is what.

- Add a BSD license (you can copy & adapt from Zeek's COPYING)

- Some of the code isn't consistently formatted, I think I saw some different indentation styles.

rsmmr commented:
This looks good!  I left some comment/feedback. Sorry for being a bit nitpicky, looking at it mostly from the perspective of fitting it into Zeek as a reusable library.

rsmmr commented:
We have another challenge here: The AC code is LGPL, which means we can't include it into BSD code.  We could make paraglob LGPL, too, but then we'd still have the problem when pulling it into Zeek.

0xekez commented:
Really appreciate you taking the time to give a detailed review. I've still got a lot to learn, so the more nitpicky the better. I'll get on this tomorrow :)

Re LGPL v. BSD: I looked a while for a better AC Trie, but had trouble finding one that fit the job as well as multifast. There were some other open-source alternatives that I saw though, so I can look into those. Otherwise, is it a hard requirement that everything in Zeek has a BSD license?

rsmmr commented:
It doesn't necessarily have to have a BSD license but it needs to be compatible with Zeek's one. It may depend on how exactly we'll integrate this into Zeek. If we licensed paraglob itself as LGPL and then compiled it into a shared library for Zeek to include, I believe that would work. However, it would also mean that everybody would need to have that shared library around to run Zeek, which doesn't seem ideal for something that's pretty narrow functionality. I'd prefer to just compile paraglob into the Zeek code --- but with LGPL we can't do that afaik.

0xekez committed:
Allow use of MultiFast under BSD

email with multifast author:
Hi Zake,

Sure. You can use MultiFast in the Zeek project repos.

THE MULTIFAST PROJECT (http://multifast.sourceforge.net/) CAN BE USED IN THE ZEEK REPOSITORY PROJECTS (https://github.com/zeek/) UNDER BSD LICENCE.

Cheers
--
Kamiar

Final Makefile change for Zeek integration
